Refactor setting the dynamic range limit in CALayers

squelart opened the pull request:
#### 084a860905a9fd5825ccfcdec0a95301dbacd86c
<pre>
Refactor setting the dynamic range limit in CALayers
<a href="https://bugs.webkit.org/show_bug.cgi?id=289500">https://bugs.webkit.org/show_bug.cgi?id=289500</a>
<a href="https://rdar.apple.com/146698266">rdar://146698266</a>

Reviewed by Mike Wyrzykowski.

Refactor repeated code (to set a dynamic range limit into one or more CALayers)
into a single implementation.

* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
Added PlatformDynamicRangeLimitCocoa.mm, and ensure header
PlatformDynamicRangeLimitCocoa.h is accessible from WebKit.

* Source/WebCore/platform/graphics/ca/cocoa/PlatformDynamicRangeLimitCocoa.h:
(WebCore::setLayerDynamicRangeLimit):
(WebCore::platformDynamicRangeLimitString): Deleted.
* Source/WebCore/platform/graphics/ca/cocoa/PlatformDynamicRangeLimitCocoa.mm: Added.
(WebCore::layerDynamicRangeLimitSetter):
Hide implementation of setting the dynamic range limit into CALayers.
`layerDynamicRangeLimitSetter` is more efficient when applying the same limit to
multiple layers.

* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::setPlatformDynamicRangeLimit):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::ensureLayer):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::setPlatformDynamicRangeLimit):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::layersAreInitialized):
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::setPlatformDynamicRangeLimit):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::setDynamicRangeLimitRecursive):
(WebKit::setDynamicRangeLimit):
(WebKit::WebViewImpl::updateHDRState):
(WebKit::setEDRStrengthRecursive): Deleted.
(WebKit::setEDRStrength): Deleted.
Use the new PlatformDynamicRangeLimitCocoa functions.

* Source/WebCore/platform/graphics/cg/UTIRegistry.mm:
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
* Source/WebCore/platform/graphics/mac/controls/SwitchThumbMac.mm:
Added missing includes due to shuffling of files in unified build.

Canonical link: <a href="https://commits.webkit.org/291990@main">https://commits.webkit.org/291990@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/36e506482fe8e4bf9af6b3ae87581ad7765c2b12

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94447 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/14039 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3819 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99466 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44974 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/96497 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14335 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22467 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72077 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29400 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97449 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10657 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85274 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52407 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10351 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2966 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44289 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80597 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/3067 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101510 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21501 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15685 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81077 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21750 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81294 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80452 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20107 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24999 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2376 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14726 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21478 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26638 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/21162 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24622 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22903 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/platform/graphics/ca/cocoa/PlatformDynamicRangeLimitCocoa.mm:
```diff
@@ -0,0 +1,59 @@
+LayerDynamicRangeLimitSetter layerDynamicRangeLimitSetter(PlatformDynamicRangeLimit platformDynamicRangeLimit)
+{
+#if HAVE(SUPPORT_HDR_DISPLAY_APIS)
+    if ([CALayer respondsToSelector:@selector(setPreferredDynamicRange:)]) {
```
```suggestion
    if ([CALayer instancesRespondToSelector:@selector(setPreferredDynamicRange:)]) {
```

pxlcoder commented on Source/WebCore/platform/graphics/ca/cocoa/PlatformDynamicRangeLimitCocoa.mm:
```diff
@@ -0,0 +1,59 @@
+
+        constexpr auto betweenStandardAndConstrainedHigh = (PlatformDynamicRangeLimit::standard().value() + PlatformDynamicRangeLimit::constrainedHigh().value()) / 2;
+        if (platformDynamicRangeLimit.value() < betweenStandardAndConstrainedHigh)
+            return [](CALayer* layer) { [layer setPreferredDynamicRange:WebKitCADynamicRangeStandard]; };
```
```suggestion
            return [](CALayer *layer) { [layer setPreferredDynamicRange:WebKitCADynamicRangeStandard]; };
```

Same comment applies to all Obj-C pointer usage.

pxlcoder commented on Source/WebCore/platform/graphics/ca/cocoa/PlatformDynamicRangeLimitCocoa.mm:
```diff
@@ -0,0 +1,59 @@
+        // FIXME: Unstage, see follow-up to rdar://145750574
+        static CADynamicRange const WebKitCADynamicRangeStandard = (id) CFSTR ("standard");
+        static CADynamicRange const WebKitCADynamicRangeConstrainedHigh = (id) CFSTR ("constrainedHigh");
+        static CADynamicRange const WebKitCADynamicRangeHigh = (id) CFSTR ("high");
```
`CADynamicRange` is an `NSString` typedef, do we really need `CFSTR` + the cast?

pxlcoder commented on Source/WebCore/platform/graphics/ca/cocoa/PlatformDynamicRangeLimitCocoa.mm:
```diff
@@ -0,0 +1,59 @@
+
+#if HAVE(SUPPORT_HDR_DISPLAY_APIS)
+#import <pal/spi/cocoa/QuartzCoreSPI.h>
+#endif // HAVE(SUPPORT_HDR_DISPLAY_APIS)
```
The import doesn't really need to be conditionalized.

pxlcoder commented on Source/WebCore/platform/graphics/ca/cocoa/PlatformDynamicRangeLimitCocoa.h:
```diff
@@ -29,31 +29,20 @@
-    static CADynamicRange const WebKitCADynamicRangeStandard = (id) CFSTR ("standard");
-    static CADynamicRange const WebKitCADynamicRangeConstrainedHigh = (id) CFSTR ("constrainedHigh");
-    static CADynamicRange const WebKitCADynamicRangeHigh = (id) CFSTR ("high");
+using LayerDynamicRangeLimitSetter = void(*)(CALayer*);
```
```suggestion
using LayerDynamicRangeLimitSetter = void(*)(CALayer *);
```

pxlcoder commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.mm:
```diff
@@ -448,12 +448,8 @@ void getSupportedTypes(HashSet<String>& types) const final
-            [sampleBufferDisplayLayer->rootLayer() setPreferredDynamicRange:platformDynamicRangeLimitString(player->platformDynamicRangeLimit())];
-    }
-#endif // HAVE(SUPPORT_HDR_DISPLAY_APIS)
+    if (auto player = m_player.get())
```
```suggestion
    if (RefPtr player = m_player.get())
```

pxlcoder commented on Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.mm:
```diff
@@ -1202,14 +1198,10 @@ static inline CGAffineTransform videoTransformationMatrix(VideoFrame& videoFrame
-        if (auto* rootLayer = sampleBufferDisplayLayer->rootLayer(); rootLayer && [rootLayer respondsToSelector:@selector(setPreferredDynamicRange:)])
-            [rootLayer setPreferredDynamicRange:platformDynamicRangeLimitString(platformDynamicRangeLimit)];
+        if (auto* rootLayer = sampleBufferDisplayLayer->rootLayer())
+            setLayerDynamicRangeLimit(rootLayer, platformDynamicRangeLimit);
```
```suggestion
        if (RetainPtr rootLayer = sampleBufferDisplayLayer->rootLayer())
            setLayerDynamicRangeLimit(rootLayer.get(), platformDynamicRangeLimit);
```

mwyrzykowski commented on Source/WebCore/platform/graphics/ca/cocoa/PlatformDynamicRangeLimitCocoa.mm:
```diff
@@ -0,0 +1,57 @@
+
+        constexpr auto betweenConstrainedHighAndHigh = (PlatformDynamicRangeLimit::constrainedHigh().value() + PlatformDynamicRangeLimit::noLimit().value()) / 2;
+        if (platformDynamicRangeLimit.value() < betweenConstrainedHighAndHigh)
+            return [](CALayer *layer) { [layer setPreferredDynamicRange:WebKitCADynamicRangeConstrainedHigh]; };
```
is `setPreferredDynamicRange` animatable?

mwyrzykowski commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -1254,36 +1255,29 @@ static bool isInRecoveryOS()
+    if ([layer wantsExtendedDynamicRangeContent]) {
     ALLOW_DEPRECATED_DECLARATIONS_END
-        if (animate) {
-            CASpringAnimation* animation = [[CASpringAnimation alloc] initWithPerceptualDuration:3.f bounce:0];
```
unsure if we need to handle animation explicitly or if this is handled by setPreferredDynamicRange - https://github.com/WebKit/WebKit/pull/42220/files#r1990288946

squelart replied:
I believe `setPreferredDynamicRange` should automatically animate in a standard manner when inside a `CATransaction`.

squelart committed:
Refactor setting the dynamic range limit in CALayers
https://bugs.webkit.org/show_bug.cgi?id=289500
rdar://146698266

Reviewed by Mike Wyrzykowski.

Refactor repeated code (to set a dynamic range limit into one or more CALayers)
into a single implementation.

* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
Added PlatformDynamicRangeLimitCocoa.mm, and ensure header
PlatformDynamicRangeLimitCocoa.h is accessible from WebKit.

* Source/WebCore/platform/graphics/ca/cocoa/PlatformDynamicRangeLimitCocoa.h:
(WebCore::setLayerDynamicRangeLimit):
(WebCore::platformDynamicRangeLimitString): Deleted.
* Source/WebCore/platform/graphics/ca/cocoa/PlatformDynamicRangeLimitCocoa.mm: Added.
(WebCore::layerDynamicRangeLimitSetter):
Hide implementation of setting the dynamic range limit into CALayers.
`layerDynamicRangeLimitSetter` is more efficient when applying the same limit to
multiple layers.

* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::setPlatformDynamicRangeLimit):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::ensureLayer):
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::setPlatformDynamicRangeLimit):
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaStreamAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::layersAreInitialized):
(WebCore::MediaPlayerPrivateMediaStreamAVFObjC::setPlatformDynamicRangeLimit):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::setDynamicRangeLimitRecursive):
(WebKit::setDynamicRangeLimit):
(WebKit::WebViewImpl::updateHDRState):
(WebKit::setEDRStrengthRecursive): Deleted.
(WebKit::setEDRStrength): Deleted.
Use the new PlatformDynamicRangeLimitCocoa functions.

* Source/WebCore/platform/graphics/cg/UTIRegistry.mm:
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
* Source/WebCore/platform/graphics/mac/controls/SwitchThumbMac.mm:
Added missing includes due to shuffling of files in unified build.

Canonical link: https://commits.webkit.org/291990@main
